src/event/base.rs

```diff
@@ -0,0 +1,197 @@
+
+pub type EventCallbackFn = extern "C" fn(EvutilSocket, EventCallbackFlags, EventCallbackCtx);
+pub type EventCallbackCtx = *mut c_void;
+pub type EventCallbackFlags = c_short;
+
+fn to_timeval(duration: Duration) -> libevent_sys::timeval {
+    libevent_sys::timeval {
+        tv_sec: duration.as_secs() as _,
+        tv_usec: duration.subsec_micros() as _,
+    }
+}
+
+pub struct EventBase {
+    base: *mut libevent_sys::event_base,
+}
+
+/// The handle that abstracts over libevent's API in Rust.
+impl EventBase {
+    pub fn new() -> Result<Self, io::Error> {
+        let base = unsafe { libevent_sys::event_base_new() };
+
+        if base.is_null() {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                "Failed to create libevent base",
+            ));
+        }
+
+        Ok(EventBase { base })
+    }
+
+    pub fn as_inner(&self) -> *const libevent_sys::event_base {
+        self.base as *const libevent_sys::event_base
+    }
+
+    pub fn as_inner_mut(&mut self) -> *mut libevent_sys::event_base {
+        self.base
+    }
+
+    pub fn loop_(&self, flags: LoopFlags) -> ExitReason {
+        let exit_code =
+            unsafe { libevent_sys::event_base_loop(self.base, flags.bits() as i32) as i32 };
+
+        match exit_code {
+            0 => {
+                unsafe {
+                    // Technically mutually-exclusive from `got_break`, but
+                    // the check in `event_base_loop` comes first, so the logic
+                    // here matches.
+                    if libevent_sys::event_base_got_exit(self.base) != 0i32 {
+                        ExitReason::GotExit
+                    } else if libevent_sys::event_base_got_break(self.base) != 0i32 {
+                        ExitReason::GotBreak
+                    } else {
+                        // TODO: This should match flags for `EVLOOP_ONCE`, `_NONBLOCK`, etc.
+                        ExitReason::Unknown { flags, exit_code }
+                    }
+                }
+            }
+            -1 => ExitReason::Error,
+            1 => ExitReason::NoPendingEvents,
+            _ => ExitReason::Unknown { flags, exit_code },
+        }
+    }
+
+    pub fn loopexit(&self, timeout: Duration) -> i32 {
+        let tv = to_timeval(timeout);
+        unsafe {
+            let tv_cast = &tv as *const libevent_sys::timeval;
+            libevent_sys::event_base_loopexit(self.base, tv_cast) as i32
+        }
+    }
+
+    pub fn loopbreak(&self) -> i32 {
+        unsafe { libevent_sys::event_base_loopbreak(self.base) as i32 }
+    }
+
+    pub fn loopcontinue(&self) -> i32 {
+        unsafe { libevent_sys::event_base_loopcontinue(self.base) as i32 }
+    }
+
+    pub fn event_new(
+        &mut self,
+        fd: Option<EvutilSocket>,
+        flags: EventFlags,
+        callback: EventCallbackFn,
+        callback_ctx: Option<EventCallbackCtx>,
+    ) -> EventHandle {
+        let fd: EvutilSocket = if let Some(fd_) = fd {
```

Comment:
Is the `_` suffix the idiomatic naming convention?  Since the original `fd` is being shadowed regardless seems like it wouldn't really matter, but curious as to the rationale.  Applies elsewhere too.

Reply:
Force of habit when funneling scoped values like this. Technically this could be simplified even further with `unwrap_or`, but I feel like that might obscure the magic of the `-1` value. Will nuke the underscore in any case.